pkg/plugin/cilium/cilium_linux.go

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+// Package cilium contains the Retina Cilium plugin. It uses unix socket to get events from cilium and decode them to flow objects.
+package cilium
+
+import (
+	"context"
+	"encoding/gob"
+	"errors"
+	"io"
+	"net"
+	"time"
+
+	v1 "github.com/cilium/cilium/pkg/hubble/api/v1"
+	"github.com/cilium/cilium/pkg/monitor/payload"
+	kcfg "github.com/microsoft/retina/pkg/config"
+	"github.com/microsoft/retina/pkg/enricher"
+	"github.com/microsoft/retina/pkg/log"
+	"github.com/microsoft/retina/pkg/plugin/api"
+	"go.uber.org/zap"
+)
+
+const (
+	MonitorSockPath1_2 = "/var/run/cilium/monitor1_2.sock"
+	defaultAttempts    = 10
+	defaultRetryDelay  = 12 * time.Second
+)
+
+func New(cfg *kcfg.Config) api.Plugin {
+	return &cilium{
+		cfg:         cfg,
+		l:           log.Logger().Named(string(Name)),
+		retryDelay:  defaultRetryDelay,
+		maxAttempts: defaultAttempts,
+		sockPath:    MonitorSockPath1_2,
+	}
+}
+
+func (c *cilium) Name() string {
+	return string(Name)
+}
+
+func (c *cilium) Generate(_ context.Context) error {
+	return nil
+}
+
+func (c *cilium) Compile(_ context.Context) error {
+	return nil
+}
+
+func (c *cilium) Init() error {
+	c.p = &parser{
+		l: c.l,
+	}
+	err := c.p.Init()
+	if err != nil {
+		c.l.Error("Failed to initialize cilium parser", zap.Error(err))
+		return err
+	}
+	c.l.Info("Initialized cilium plugin")
+	return nil
+}
+
+func (c *cilium) Start(ctx context.Context) error {
+	if c.cfg.EnablePodLevel {
+		if enricher.IsInitialized() {
+			c.enricher = enricher.Instance()
+		} else {
+			c.l.Warn("retina enricher is not initialized")
+		}
+	}
+
+	// Connect and monitor loop
+	err := c.connect(ctx)
+	if err != nil {
+		c.l.Error("Error while connecting and decoding cilium events", zap.Error(err))
+		return err
+	}
+	go func() {
+		c.monitorLoop(ctx)
+	}()
+
+	<-ctx.Done()
+	return nil
+}
+
+func (c *cilium) Stop() error {
+	c.l.Info("Stopped cilium plugin")
+	return nil
+}
+
+func (c *cilium) SetupChannel(ch chan *v1.Event) error {
+	c.externalChannel = ch
+	return nil
+}
+
+// Create a connection to the cilium unix socket to monitor events
+func (c *cilium) connect(ctx context.Context) error {
+	// Start the cilium monitor
+	for attempt := 0; attempt < c.maxAttempts; attempt++ {
+		select {
+		case <-ctx.Done(): // Cancelled or done
+			//nolint:wrapcheck // dont wrap error since it would not provide more context
+			return ctx.Err()
+		default:
+			conn, err := net.Dial("unix", c.sockPath)
+			if err != nil {
+				c.connection = nil
+				c.l.Error("Failed to connect to cilium monitor", zap.Error(err))
+				time.Sleep(c.retryDelay)
+				continue
+			}
+			c.l.Info("Connected to cilium monitor")
+			c.connection = conn
+			return nil
+		}
+	}
+	return nil
+}
+
+// monitor events from uds connection
+func (c *cilium) monitorLoop(ctx context.Context) {
```

Comment:
I don't think `monitor` should be trying to handle connection failures or manipulating the connection in any way. 

There are three components at play here:
- connection manager
- reader
- parser

The plugin should have a controller that synchronizes the three components:
- try setting up connection
- try reading data
- try parsing data

The individual components should decide which error is "warn" and needs retry and which error should be reported. The controller, not the components, should decide course of action. For ex: if reader reports connection error, controller should retry connection and start reader again. If connection manager reports error, that is a fatal error and controller should report back to plugin manager.

Above pattern would establish clear responsibilities and easy to follow/maintain code.